### PR TITLE
Implement the Cargo half of pipelined compilation (take 2)

### LINKED ISSUE #6660: Implement "pipelined" rustc compilation
One possible feature we've talked about for a long time but never got around to implementing is the concept of pipelined rustc compilation. Currently today let's say that we have a crate A and a crate B that depends on A. Let's also say we're both compiling rlibs. When compiling this project Cargo will compile A first and then wait for it to finish completely before starting to compile B.

In reality, though, the compiler doesn't need the full compilation results of A to start B. Instead the compiler only needs *metadata* from A to start compiling B. Ideally Cargo would start compiling B as soon as A's metadata is ready to go.

This idea of pipelining rustc and starting rustc sooner doesn't reduce the overal work being done on each build, but it does in theory greatly increase the parallelism of the build as we can spawn rustc faster and keep all of a machine's cores warm doing Rust compilation. This is expected to have even bigger wins in release mode where post-metadata work in the compiler often takes quite some time. Furthermore incremental release builds should see *huge* wins because during incremental rebuilds of a chain of crates you can keep all cores busy instead of just dealing with one crate at a time.

There's three main parts of this implementation that need to happen:

* [ ] First, the compiler needs to be fixed to actually accept metadata as input for dependencies when producing an rlib. I believe currently it requires all inputs to be rlibs, but it shouldn't be too hard to fix this. (Tracking issue for rustc changes is https://github.com/rust-lang/rust/issues/58465)
* [ ] Second, the compiler will need to send a signal to Cargo as soon as metadata is produced. Cargo will instruct rustc to produce both metadata and an rlib, and once Cargo gets a signal metadata is ready then rustc will go on continuing to produce an rlib and Cargo will start subsequent rustc invocations that it can. Note that when I say "signal" here I don't really mean unix signals, but rather something like a message over a TCP socket or something like that.
* [ ] Finally Cargo will need to be refactored to listen for these signals, fork off more work in parallel, and properly synchronize the results of everything. This is where I suspect the bulk of the work will be happening (hence the issue in this repository).

In the ideal world the compiler would *also* wait just before linking for Cargo to let it know that all dependencies are ready. That's somewhat difficult, however, so I think it's probably best to start out incrementally and simply say that Cargo doesn't start a compilation that requires linking until all dependencies are finished (as it does today).

I've talked with @ehuss about possibly implementing this as well as the compiler team about the idea, but I think this is a small enough chunk of work (although certainly not trivial) to be done in the near future!

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit starts to lay the groundwork for #6660 where Cargo will
invoke rustc in a "pipelined" fashion. The goal here is to execute one
command to produce both an `*.rmeta` file as well as an `*.rlib` file
for candidate compilations. In that case if another rlib depends on that
compilation, then it can start as soon as the `*.rmeta` is ready and not
have to wait for the `*.rlib` compilation.

Initially attempted in #6864 with a pretty invasive refactoring this
iteration is much more lightweight and fits much more cleanly into
Cargo's backend. The approach taken here is to update the
`DependencyQueue` structure to carry a piece of data on each dependency
edge. This edge information represents the artifact that one node
requires from another, and then we a node has no outgoing edges it's
ready to build.

A dependency on a metadata file is modeled as just that, a dependency on
just the metadata and not the full build itself. Most of cargo's backend
doesn't really need to know about this edge information so it's
basically just calculated as we insert nodes into the `DependencyQueue`.
Once that's all in place it's just a few pieces here and there to
identify compilations that *can* be pipelined and then they're wired up
to depend on the rmeta file instead of the rlib file.

Closes #6660

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @ehuss 

So I've got one open question on this implementation which is included as a failing test here. The general question is what do we want to happen in this scenario:

* In the first compile, a user builds an rlib.
* In the next compile, the user builds another rlib that depends on the first.

The test executes this as `cargo build -p dependency` followed by `cargo build -p parent`. In this scenario it's currently buggy because the second compilation assumes the first produced metadata, but we currently try to avoid producing metadata where possible. The options to solve this are:

1. Always produce metadata files for rlibs, whether or not the metadata will be consumed during pipelining
2. In the second compile recompile the dependency rlib to produce metadata. (we'd handle this by hashing whether metadata is produced into the fingerprint)

Option (1) would take up a bit more disk space while option (2) may take longer, but this seems unlikely to really come up in practice all that often. I'd slightly lean towards option (1), but I'm curious what you'd think as well?

**Comment by Eh2406** [comment]
Are these the same metadata files need for 'cargo check' to use the results of 'cargo build'?

**Comment by alexcrichton** [comment]
I think in theory they can be compatible, but this doesn't attempt to set things up so `cargo check` can be shared with `cargo build`, but we can perhaps tackle that not too difficultly in the future!

**Comment by ehuss** [comment]
> I'd slightly lean towards option (1), but I'm curious what you'd think as well?

Either option sounds fine to me. It doesn't seem like a huge amount of extra disk space.

**Comment by alexcrichton** [comment]
Ok, I've implemented that strategy since it's easier!

**Comment by ehuss** [comment]
Looks like a bunch of tests need to be updated now that `--emit` has been expanded.

**Comment by alexcrichton** [comment]
Oops, updated that

**Comment by ehuss** [comment]
Looks good to me.  r=me if you're ready to go.  I'm not sure if you want to wait for the rustc changes or not.  I figure worst case if things don't work out, this can be reverted?

**Comment by alexcrichton** [comment]
I think it makes sense to wait for an end-to-end story to prove out that it's working. To that end I've pushed up a few more commits which I believe fully implements pipelined compilation in Cargo given the support in upstream rustc for `-Z emit-directives`. Some caveats apply though:

* This'll fail CI since we don't have a nightly which supports `emit-directives`
* This'll have very little benefit for pipelined compilation since rustc doesn't emit the directive until the very end of compilation. The compiler is being changed in https://github.com/rust-lang/rust/pull/60385 to emit the directive much sooner
* Still requires opt in, for example through `CARGO_BUILD_PIPELINING=1`

**Comment by alexcrichton** [comment]
Ok I think that covers everything and I've verified it works with tonights nightly, so this should be good to go!

**Comment by alexcrichton** [comment]
@ehuss did you want to give this a final once-over before merging for the commits I added after the initial review?

**Comment by ehuss** [comment]
I'm still working through this.  I don't see any noticeable improvement in total compilation time with the various projects I've tried.  Is there a particular project type or concurrency level that would see a benefit?  I'm concerned about moving forward if it ultimately doesn't help.

I've been instrumenting this a little.  I've been building some graphs of the queue sizes, which shows that the initial phase is kept more "full", but that drops off quickly.  I've also been recording how much faster the rmeta files are emitted. On average, the rmeta file is emitted at about the 70% mark of the total crate build time (ranging from 30% to 98%). I'm not sure if that intuitively seems reasonable. I believe it is currently emitted just before codegen?

I'm going to continue instrumenting and try to get a better picture.  Would like to get your perspective.

**Comment by alexcrichton** [comment]
Heh a good point on the investigations! My plan was moreso "land it and make a post asking for users to test it out" and debug from there :)

I was also slightly dismayed at the lack of benefit on Cargo itself in both release and debug builds. I suspect that this is not going to be as large of a win as I had previously hoped. That being said, I'm still hopeful we can get testers on internals to help show some other data.

This also isn't necessarily a win in all cases. The whole point of pipelining is to take advantage of otherwise idle parallelism, and we can't pipeline literally everything and what's here is a coarse approximation of the ideal. Some ingredients necessary to reap benefits from this PR:

* You'll at least have to compile more than one crate. Incremental builds that build the top crate will see no benefit.
* You'll probably need to be compiling in release mode to get the most benefit. Otherwise codegen is typically quick enough that it won't really feel all that different. (AKA if codegen is "fast enough" in debug mode pipelining won't actually start anything that much sooner).
* There may not be a ton of benefit in full crate graph builds. There's lots of edges that still require full synchronization of build process, such as:
  * Procedural macros (aka serde_derive) have to be entirely finished before proceeding
  * Build scripts (aka building large chunks of C code) must entirely finish as well

Given all that I suspect that the scenario with the biggest win will be incremental release mode builds where a crate a few layers deep in the graph was modified. Incremental builds in debug mode may fit this same profile but will likely not feel all that much faster.

FWIW though I'm personally pretty happy with how low the complexity level turned out for this PR, so I think it'd be fine to land with the intention of backing it out if it doesn't pan out. That being said though I can't think of a use case where this *regresses* something, so I'm not sure there's a ton of downside to not landing this. (that's exclusively given the low complexity budget spend I'm perceiving here)

**Comment by stuhood** [comment]
> I was also slightly dismayed at the lack of benefit on Cargo itself in both release and debug builds. I suspect that this is not going to be as large of a win as I had previously hoped. That being said, I'm still hopeful we can get testers on internals to help show some other data.

From simulations we've done with pipelining/"outlining", the benefits only begin to show on beefier machines with larger graphs. If you don't have enough total work, or if you don't have enough cores to run more of the unblocked work in parallel, then the outlining is overhead (but hopefully a very small amount of overhead).

A case where outlining helps even for smaller workspaces is the "deep path" case. In a workspace with a chain of dependencies, if you edit a few deps away from a test/binary, the "path" between them can be compiled in parallel, whereas without outlining it would be sequential.

File: src/cargo/core/compiler/mod.rs [hunk 1]
```diff
@@ -789,6 +805,11 @@ fn build_base_args<'a, 'cfg>(
 
     if unit.mode.is_check() {
         cmd.arg("--emit=dep-info,metadata");
+    } else if !unit.target.requires_upstream_objects() {
+        // Always produce metdata files for rlib outputs. Metadata may be used
+        // in this session for a pipelined compilation, or it may be used in a
+        // future Cargo session as part of a pipelined compile.c
```
Comment (ehuss) [inline src/cargo/core/compiler/mod.rs]:
```suggestion
        // future Cargo session as part of a pipelined compile.
```

File: src/cargo/core/compiler/mod.rs [hunk 2]
```diff
@@ -1046,32 +1070,88 @@ impl Kind {
...
+    target: &Target,
+    extract_rendered_errors: bool,
+    look_for_metadata_directive: bool,
+) -> CargoResult<()> {
+    // We primarily want to use this function to process JSON messages from
+    // rustc. The compiler should always print one JSON message per line, and
+    // otherwise it may have other output intermingled (think RUST_LOG or
+    // something like that), so skip over everything that doesn't look like a
+    // JSON message.
+    if !line.starts_with('{') {
```
Comment (ehuss) [inline src/cargo/core/compiler/mod.rs]: How confident are you that non-JSON output will never start with `{` (particularly with human error-format)? I haven't been able to think of a way to confuse it, but I'm concerned that this could inadvertently interpret a line as json.  The most likely case would be RUSTC_LOG debug output, though that might be a bit far fetched.

File: src/cargo/core/compiler/mod.rs [hunk 3]
```diff
@@ -1046,32 +1070,88 @@ impl Kind {
...
+        .map_err(|_| internal(&format!("compiler produced invalid json: `{}`", line)))?;
+
+    // In some modes of compilation Cargo switches the compiler to JSON mode but
+    // the user didn't request that so we still want to print pretty rustc
+    // colorized errors. In those cases (`extract_rendered_errors`) we take a
+    // look at the JSON blob we go, see if it's a relevant diagnostics, and if
+    // so forward just that diagnostic for us to print.
+    if extract_rendered_errors {
+        #[derive(serde::Deserialize)]
+        struct CompilerError {
```
Comment (ehuss) [inline src/cargo/core/compiler/mod.rs]: Can this be renamed `CompilerMessage`, since these aren't always "errors"?

File: src/cargo/core/compiler/mod.rs [hunk 4]
```diff
@@ -1046,32 +1070,88 @@ impl Kind {
...
+    // And failing all that above we should have a legitimate JSON diagnostic
+    // from the compiler, so wrap it in an external Cargo JSON message
+    // indicating which package it came from and then emit it.
+    let msg = machine_message::emit(&machine_message::FromCompiler {
+        package_id,
+        target,
+        message: compiler_message,
+    });
+
+    // Switch json lines from rustc/rustdoc that appear on stderr to instead.
```
Comment (ehuss) [inline src/cargo/core/compiler/mod.rs]:
```suggestion
    // Switch json lines from rustc/rustdoc that appear on stderr to stdout instead.
```
?

File: src/cargo/util/machine_message.rs [hunk 5]
```diff
@@ -10,11 +10,11 @@ pub trait Message: ser::Serialize {
     fn reason(&self) -> &str;
 }
 
-pub fn emit<T: Message>(t: &T) {
+pub fn emit<T: Message>(t: &T) -> String {
```
Comment (ehuss) [inline src/cargo/util/machine_message.rs]: Can this be renamed, now that it no longer emits?  And maybe make it a `Message` trait method?

**Comment by ehuss** [comment]
Sounds reasonable.

I'd like to add some timing data to the debug logs, I'll follow up with a separate PR with some ideas.

I think this will address some other issues, like #6674 and #6119.

> but it turns out that we unconditionally capture all output from the compiler/rustdoc today, no exceptions

This hasn't been true since #6289.

**Comment by ehuss** [comment]
May also want to wait for https://github.com/rust-lang/rust/pull/60464 which is currently running with bors.

**Comment by matthiaskrgr** [comment]
I was trying this out (merged lastest master into it before trying out) and it seems that sometimes the progress bar is eaten by warning messages where it previously was not (unless the bug was introduced by some other change in the meantime)

cargo 1.36.0-nightly (6be12653d 2019-04-19):
````
warning: the item `humantime` is imported redundantly
  --> src/internal/filter/time.rs:12:13
   |
12 |         use humantime;
   |             ^^^^^^^^^
   |
   = note: #[warn(unused_imports)] on by default

    Building [======================================================>  ] 47/48: fd(bin)
````

this branch:

````
warning: the item `humantime` is imported redundantly
  --> src/internal/filter/time.rs:12:13
   |
12 |         use humantime;
   |             ^^^^^^^^^
   |
   = note: #[warn(unused_imports)] on by default

````
(no progress bar shown when building `fd(bin)` target.

I was `cargo build --release`ing fd: https://github.com/sharkdp/fd

**Comment by matthiaskrgr** [comment]
This started with f7bf43a0494ac8c4913fe10b48176ada8749efeb / "Implement the Cargo half of pipelined compilation"
and is reproducible without `CARGO_BUILD_PIPELINING=true`

**Comment by ehuss** [comment]
> sometimes the progress bar is eaten

Ah, thanks for catching that!

I'm not thinking very clearly right now, but I'm pretty sure it is due to the anti-flicker code for the progress bar.  It doesn't re-display the progress bar if it thinks that it hasn't changed.  I think the solution is in `drain_the_queue` the `Message::Stderr` handler needs to call `self.progress.clear();` before printing.  I don't remember exactly why it doesn't already do that.

**Comment by ehuss** [comment]
I'm getting a strange error trying to build `mdbook`.  I can only repro it on my mac, so something is very strange.

```
error: crate `syn` required to be available in rlib format, but was not found in this form

error: aborting due to previous error

error: Could not compile `pest_derive`.
```

It's a bit of a race condition, because once `syn` finishes building, the `-L` flag picks it up.

I'm wondering if `Artifact::All` isn't getting wired up correctly.

I'll try to investigate more later.

**Comment by ehuss** [comment]
I seem to be missing something fundamental.  How are the "All" edges getting added for transitive deps?  For example:

```
a (binary)
└── b (lib)
    └── c (lib)
```

Here, there is an "All" edge from `a → b`, and a Metadata edge from `b → c`.  Shouldn't there also be an All edge from `a → c`?  If `c` takes a long time to do codegen, then `a` will start before `c` is finished, and fail.

Is there an easy way to force codegen to take a very long time?  I want to force some of these delays.

Also, fingerprinting seems to be broken.  Rebuilding a project seems to frequently need rebuilding various dependencies.

**Comment by ehuss** [comment]
Another issue:  `cargo clean -p pkg` does not remove the `.rmeta` file generated by the package.  I think [around here](https://github.com/rust-lang/cargo/blob/759b6161a328db1d4863139e90875308ecd25a75/src/cargo/ops/cargo_clean.rs#L118) it might need to check the `flavor`, but that seems rather hackish.  It might be good if `outputs` could somehow return the `.rmeta` file separately, since there are other parts of the code that assume it's an actual list of all outputs (fingerprints, collision detection, etc.).  Not sure what the full consequences are.

**Comment by alexcrichton** [comment]
Er sorry was distracted yesterday and didn't get a lot done!

> I'd like to add some timing data to the debug logs, I'll follow up with a separate PR with some ideas.

You're also more than welcome to just push them here if you'd like, it may take a minute to solve the compilation issue you found.

> >  but it turns out that we unconditionally capture all output from the compiler/rustdoc today, no exceptions
>
> This hasn't been true since #6289.

Ah yeah you got what I meant in your edit, but I meant "capture" in the sense of redirecting fds rather than literally holding it in memory!

> May also want to wait for rust-lang/rust#60464 which is currently running with bors.

Agreed! It's now landed too but hasn't made its way to nightly just yet. Let's hold off on merging until that happens (presumably tonight) and I can fixup the routing done here.

>  I think the solution is in drain_the_queue the Message::Stderr handler needs to call self.progress.clear(); before printing. I don't remember exactly why it doesn't already do that.

Hm internally `print_ansi` should do the clearing. I'll investigate what's happening here.

> I seem to be missing something fundamental. How are the "All" edges getting added for transitive deps? 
>
> ...
> 
> Is there an easy way to force codegen to take a very long time? I want to force some of these delays.
>
> Also, fingerprinting seems to be broken. Rebuilding a project seems to frequently need rebuilding various dependencies.

I think I mistakenly thought they got wired up automatically. In the previous iteration this naturally fell out, but thinking about this version I believe you're right in that it's not wired up correctly currently. I'll work on fixing that.

For long codegen, I'd recommend a macro which expands to thousands of `println!` statements or something like that.

For rebuilding, I'll try to reproduce and investigate

> Another issue: cargo clean -p pkg does not remove the .rmeta file generated by the package

I'll try to look into this too

**Comment by alexcrichton** [comment]
For an example of something that takes a very long time in codegen:

```rust
use std::io::Write;

macro_rules! a {
    (1) => ({
        let x = |a: u32| a + 3;
        assert_eq!(x(3), 6);
    });
    (2) => ({a!(1); a!(1);});
    (3) => ({a!(2); a!(2);});
    (4) => ({a!(3); a!(3);});
    (5) => ({a!(4); a!(4);});
    (6) => ({a!(5); a!(5);});
    (7) => ({a!(6); a!(6);});
    (8) => ({a!(7); a!(7);});
    (9) => ({a!(8); a!(8);});
    (10) => ({a!(9); a!(9);});
    (11) => ({a!(10); a!(10);});
    (12) => ({a!(11); a!(11);});
    (13) => ({a!(12); a!(12);});
    (14) => ({a!(13); a!(13);});
    (15) => ({a!(14); a!(14);});
}

fn a() {
}

pub fn foo() {
    a!(10);
}
```

(tweak the various constants as necessary for your machine to make it acceptable, or the body of the base case)

**Comment by alexcrichton** [comment]
Looks like the progress bar was a bug in the `job_queue.rs` changes unrelated to printing, it was because an accidental bug in the number of packages being compiled and jobs to do caused the progress bar to be prematurely flagged as done.

File: src/cargo/core/compiler/mod.rs (hunk 2 above)
Comment (alexcrichton) [reply src/cargo/core/compiler/mod.rs]: I'm only semi-reasonably confident on this point. It's sort of an implicit contract we have currently, but it seems reasonable to gracefully handle lines that start with `{` yet aren't JSON and just print them out anyway instead of aborting compilation.

**Comment by alexcrichton** [comment]
> Also, fingerprinting seems to be broken. Rebuilding a project seems to frequently need rebuilding various dependencies.

I think what's going on here is pretty nasty. Consider a compilation where A depends on B:

* B produces metadata
* A starts and finishes
* B finishes

Next compilation it looks like A is stale because A's mtimes are all before B's artifacts. We need to encode in fingerprinting that A's dependency on B only look at the rmeta artifact since it didn't actually consume any of the other artifacts anyway.

**Comment by alexcrichton** [comment]
Ok, I think the subsequent commits should address all issues brought up here along the way.

**Comment by matthiaskrgr** [comment]
It looks like this breaks compatibility with sccache
````
   Compiling unreachable v1.0.0
sccache: encountered fatal error
sccache: error : Failed to open file for hashing: "/home/matthias/vcs/github/fd/target/debug/deps/libvoid-f01d6c64b2fc03b4.rmeta"
sccache:  cause: Failed to open file for hashing: "/home/matthias/vcs/github/fd/target/debug/deps/libvoid-f01d6c64b2fc03b4.rmeta"
sccache:  cause: No such file or directory (os error 2)
error: Could not compile `unreachable`.
````
is this expected?

File: src/cargo/core/compiler/fingerprint.rs [hunk 6]
```diff
@@ -322,6 +323,7 @@ struct DepFingerprint {
     pkg_id: u64,
     name: String,
     public: bool,
+    only_requires_rmeta: bool,
```
Comment (ehuss) [inline src/cargo/core/compiler/fingerprint.rs]: Can you add a short doc comment explaining this field?  And remove the list above and just doc each field?

**Comment by alexcrichton** [comment]
Updated with @ehuss's comments.

@matthiaskrgr I'm not sure why, but I'm not really shooting for 100% fidelity works everywhere in this PR, that's something I hope we can fix and iterate on afterwards.

**Comment by ehuss** [comment]
Sorry, one more minor nit.  I notice the progress bar jumping backwards at times.  I think the following will fix it:

```diff
diff --git a/src/cargo/core/compiler/job_queue.rs b/src/cargo/core/compiler/job_queue.rs
index bff2d5b6..10ecf47f 100644
--- a/src/cargo/core/compiler/job_queue.rs
+++ b/src/cargo/core/compiler/job_queue.rs
@@ -320,7 +320,8 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
             // unnecessarily.
             let events: Vec<_> = self.rx.try_iter().collect();
             let events = if events.is_empty() {
-                self.show_progress(total);
+                let count = total - self.queue.len() - self.active.len() - queue.len();
+                self.show_progress(count, total);
                 vec![self.rx.recv().unwrap()]
             } else {
                 events
@@ -431,8 +432,7 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
         }
     }

-    fn show_progress(&mut self, total: usize) {
-        let count = total - self.queue.len() - self.active.len();
+    fn show_progress(&mut self, count: usize, total: usize) {
         let active_names = self
             .active
             .values()
```

Essentially when jobs enter the local `queue`, they get lost from the count until they get added to `active`.  It also made the progress bar appear to be further along than it actually was.

**Comment by alexcrichton** [comment]
Ah yeah I think I like the idea of a running finished count, so pushed that!

File: src/cargo/core/compiler/job_queue.rs [hunk 7]
```diff
@@ -350,6 +351,7 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
                         print.print(&msg)?;
                     }
                     Message::Finish(id, artifact, result) => {
+                        finished += 1;
```
Comment (ehuss) [inline src/cargo/core/compiler/job_queue.rs]: I think this needs to be moved down to the `Artifact::All` branch?

Comment (alexcrichton) [reply src/cargo/core/compiler/job_queue.rs]: Hm yes indeed

**Comment by ehuss** [comment]
:+1:

@bors r+

**Comment by mark-i-m** [comment]
So is pipelining implemented on nightly now?

**Comment by Eh2406** [comment]
No. It is on master for cargo. Every week or so there is a PR on rust to update the cargo subrepo. The next nightly after that PR is merged will have this functionality. Then the functionality still requires an opt in with the env `CARGO_BUILD_PIPELINING`.

**Comment by alexcrichton** [comment]
I've posted an [internals topic](https://internals.rust-lang.org/t/evaluating-pipelined-rustc-compilation/10199) for getting more information here
